### ISSUE TITLE
How the vocab_to_id mapping is decided in ELMo without an explicit mapping file?

### ISSUE THREAD

**Issue by scofield7419** [allenai/allennlp#3318]
**System (please complete the following information):**
- OS: Linux
- Python version: 3.6.5
- AllenNLP version: the latest from master
- PyTorch version: 1.2.0

**Question**
During my using of AllenNLP-Elmo, I'am so confused by the AllenNLP API. Following are my question statements (I'm trying to describe myself clearly).

------------

1.  For the official English ELMo pre-trained weights, people can use it to generate representation for tokens.
However, in the [official instruction](https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md), there is no explicit vocab_to_id(maybe token, or charactor) mapping file needed to be passed into the Elmo instructor.
![1](https://user-images.githubusercontent.com/18722770/66103058-12f89d80-e5e7-11e9-872f-4fa5f6c101ee.png)
Instead, it use the on-the-fly conversion: `batch_to_ids` function.

I noticed the `batch_to_ids` function converts the char to corresponding id, but I can't be informed **how the charactors be converted, and what the mapping is**.
So this makes me wonder, how the mapping work without a explicit vocab-to-id(char-level) file?

------------

2. Without knowing the mapping mechanism, brings me a big problem as following.
I'm working on a multi-lingual task, and I pre-trained the elmo on the mixture of multi-lingual snetence corpus, with [bilm-tf](https://github.com/allenai/bilm-tf).
And I obtain the three outputs:
`weights.hdf5`, `option.json`, `vocab.txt`
Then, I wanted to load this pretrained ELMo into my specific pytorch Module (with other layer on top of ELMo) by using AllenNLP.
And I come across the problems.

Initially, I consider to use the  `batch_to_ids` function to converts the sentences to char-ids, just like the [official instruction](https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md) says.
But, **how could I know** the  `batch_to_ids`  made a exact correct conversion? (since it's not on the English Scenory; instead, many other languages in which many other tokens and charactors are occurred there)

Then, I consider to use the `vocab_to_cache` to pass my own word_vocab (to avoid the char-ids, since it brings me uncertainty), which means I make the ELMo used at token-level.
But the problem is: I have to pass the token id to the model, while I was't given a token-to-id mapping.
Here is the vocab file:
![2](https://user-images.githubusercontent.com/18722770/66104285-a54e7080-e5ea-11e9-8ca5-2fc7b79e420d.jpg)

Well then, I manually assign each token with an id, by the defaut mapping rule in [the code of bilm-tf](https://github.com/Rokid/ELMo-chinese/blob/master/bilm/data.py#31).
![3](https://user-images.githubusercontent.com/18722770/66104754-2e19dc00-e5ec-11e9-9ae8-bc8835f3b4c1.png)

Note that I am not sure whether such mapping from **bilm-tf** is real correct for **allennlp-elmo**.

And I find the `<s>` has its id of 0. But this brings me another very exact problem:
Isn't the id of 0 for `Padding`? 
![4](https://user-images.githubusercontent.com/18722770/66104402-ffe7cc80-e5ea-11e9-9450-02633b152337.png)
Won't the id of 0 cause collision for `<s>` and `Padding`?

Well, so far, these problems make me feel MUCH confused!

Think about the [google-bert](https://github.com/google-research/bert), it has a unified and accordant conversion for tokens (its `BertTokenizer`), which make the conversion much more easier and clear.

---------------------

With all those regards above, I think the AllenNLP should makes the instruction or document of APIs more clear. 
Help me plz, don't make me guess.

**Comment by kernelmachine** [comment]
Here is where the mappings for the EOS/PAD tokens are defined for ELMo: https://github.com/allenai/allennlp/blob/master/allennlp/data/token_indexers/elmo_indexer.py#L27

**Comment by DeNeutoy** [comment]
@scofield7419 

The elmo model uses a byte level character representation, with a few additional symbols for <pad> and <bow>. So for characters which are utf-8 but take up multiple bytes, they will be split first into several "byte characters". Does that answer your question?

**Comment by DeNeutoy** [comment]
Closing due to inactivity
